core/src/main/scala/cats/data/Chain.scala

```diff
@@ -565,33 +565,19 @@ sealed abstract class Chain[+A] {
   /**
    * Returns the number of elements in this structure
    */
-  final def length: Long = {
-    // TODO: consider optimizing for `Chain.Wrap` case.
-    //       Some underlying seq may not need enumerating all elements to calculate its size.
-    val iter = iterator
-    var i: Long = 0
-    while (iter.hasNext) { i += 1; iter.next(); }
-    i
-  }
+  final def length: Long =
```

Comment:
actually I think what we really might want is:

```scala
def foldMap[B](fn: A => B)(implicit B: Monoid[B]): B = {
  @annotation.tailrec
  def loop(chains: List[Chain[A]], acc: B): B =
    chains match {
      case Nil => acc
      case h :: tail =>
        h match {
          case Empty => loop(tail, acc)
          case Wrap(seq) => loop(tail, B.combine(B.combineAll(seq.iterator.map(fn)), acc))
          case Singleton(a) => loop(tail, B.combine(fn(a), acc))
          case Append(l, r) => loop(l :: r :: tail, acc)
       }
}
```
That is stack safe and leverages associtivity (and we can override the foldMap in the Foldable instance).

With that, you can do: `def length: Long = foldMap(Function.const(1L))`

Or you could write a custom loop based on the same idea and make it stack safe.

Reply:
Just a note: using such `foldMap` for calculating `length` will undermine the original efforts for `Wrap` case optimization.

Reply:
...although the method `foldMap` itself is nice to have for sure.

Reply:
yeah, I mean you can make 3 specialized methods: `combineAll, foldMap, length` and each of them can have their own code.

If you just care about length right now, you can just add that one and change the `Wrap(seq)` line to be `loop(tail, acc + seq.length.toLong)`

Reply:
Just curious: is it going to be faster than a stack-safe implementation based on `Eval`? I know `Eval` uses closures, but this code also does some additional allocations of `List` items all the way while recursing. Wdyt?

Reply:
Eval also allocates lists to manage a stack. I think this will be faster since it won't also allocate and call lambdas.

Reply:
Another option with `sealed abstract private[data] trait NonEmptyBase[+A] extends Chain[A]` inherited by `Singleton` and `Wrap` (what did I say about getting sucked down rabbit-holes...)

```scala
  private def foldMapCase[B](f: NonEmptyBase[A] => B)(implicit M: Monoid[B]): B = {
    @tailrec def loop(chains: List[Chain[A]], acc: B): B = chains match {
      case Nil => acc
      case h :: t =>
        h match {
          case Empty              => loop(t, acc)
          case h: NonEmptyBase[A] => loop(t, M.combine(f(h), acc))
          case Append(l, r)       => loop(l :: r :: t, acc)
        }
    }

    loop(this :: Nil, M.empty)
  }

  final def foldMap[B](f: A => B)(implicit M: Monoid[B]): B = foldMapCase {
    case Singleton(a) => f(a)
    case Wrap(seq)    => Foldable[Seq].foldMap(seq)(f)
  }

  final def length: Long = foldMapCase {
    case Singleton(_) => 1L
    case Wrap(seq)    => seq.length.toLong
  }
```

Reply:
I wouldn't sweat it. Just copy the code. The inlined copied code will also be more efficient since there is no Long boxing.

We prefer a copy of the code if we can get efficiency wins in library code. We aren't trying to make the internals of cats as beautiful as possible. IMO, we want the API to be beautiful, then we want the library to be stack safe, then we want it to be fast, then we want it to be implemented in a beautiful fashion.

Reply:
thinking more about this, I am not sure we actually want to implement `foldMap` the way I suggested. Doing so would likely defeat the optimizations of `Monoid.combineAll`. I think the right way is just `Monoid.combineAll(toIterator.map(fn))` which is free to use internal mutability on the monoid. If we implement the way I suggested above, we force the monoid to continue to concatenate (which could make things like string concatenation quadratic vs linear).